### PR TITLE
Allow agent forwarding from Web UI and tsh

### PR DESCRIPTION
**Pull request by russjones**
**Purpose**

This PR adds the ability to forward your Teleport agent both using `tsh` and the Web UI. Note, the system agent is not forwarded in either case, we forward the Teleport agent only.

`tsh` supports for the `-A` flag which allows you the option to forward or not forward your agent. The Web UI will always attempt to forward our agent. If agent forwarding either not allowed by RBAC rules or policy on the node, the user will still be able to login, but their agent will not be forwarded.

**Implementation**

* Config for Teleport Client supports `ForwardAgent` boolean.
* Added support for `-A` in `tsh` to toggle `ForwardAgent`.
* Web UI always sets `ForwardAgent` to true.
* We always try and forward agent to proxy. We do this to lay the groundwork for the recording proxy where this will be required. This does not cause problems and is not a security risk because it's a NOP in the regular server.
* When we create a session if `ForwardAgent` is true, we request agent forwarding.

**Related Issues**

Fixes https://github.com/gravitational/teleport/issues/1000

### CODE REVIEW

**Comment by russjones** [comment]
@gravitational-jenkins retest this please

**Comment by russjones** [comment]
@gravitational-jenkins retest this please

**Comment by russjones** [comment]
@gravitational-jenkins retest this please

**Comment by klizhentas** [comment]
instead of PRing to master, can you PR to your refactoring branch for now? will be easier to review

**Comment by russjones** [comment]
@klizhentas Changed the base to `rjones/fix-agent-forward`.

File: lib/client/api.go [hunk 1]
```diff
@@ -125,6 +125,12 @@ type Config struct {
 	// against Teleport client and obtaining credentials from elsewhere.
 	SkipLocalAuth bool
 
+	// Agent is used with SkipLocalAuth and used to pass in an agent when using the Web UI.
+	Agent agent.Agent
+
+	// ForwardAgent requests that your agent be forwarded to the target node.
```
Comment (klizhentas) [inline lib/client/api.go]: `your`? :) client's is more appropriate

File: lib/client/api.go [hunk 2]
```diff
@@ -125,6 +125,12 @@ type Config struct {
 	// against Teleport client and obtaining credentials from elsewhere.
 	SkipLocalAuth bool
 
+	// Agent is used with SkipLocalAuth and used to pass in an agent when using the Web UI.
```
Comment (klizhentas) [inline lib/client/api.go]: Agent is used when `SkipLocalAuth` is `true`?

File: lib/client/api.go [hunk 3]
```diff
@@ -368,6 +374,12 @@ func NewClient(c *Config) (tc *TeleportClient, err error) {
 		if len(c.AuthMethods) == 0 {
 			return nil, trace.BadParameter("SkipLocalAuth is true but no AuthMethods provided")
 		}
+		// if we were passed in a agent in the configuration and skip local auth, use
```
Comment (klizhentas) [inline lib/client/api.go]: if the client was passed an agent

File: lib/client/api.go [hunk 4]
```diff
@@ -368,6 +374,12 @@ func NewClient(c *Config) (tc *TeleportClient, err error) {
...
+		// the passed in agent. the web ui does this as it fetches the agent from
```
Comment (klizhentas) [inline lib/client/api.go]: `web ui` is irrelevant here as the user, the client should be generic

File: lib/client/client.go [hunk 5]
```diff
@@ -249,6 +252,25 @@ func (proxy *ProxyClient) ConnectToNode(ctx context.Context, nodeAddress string,
 		}
 	}
 
+	// always try to forward agent to the proxy once (we need this check or we'll
```
Comment (klizhentas) [inline lib/client/client.go]: this logic does not belong in client, it should be done by web component.

File: tool/tsh/tsh.go [hunk 6]
```diff
@@ -167,6 +169,7 @@ func Run(args []string, underTest bool) {
 	ssh.Arg("[user@]host", "Remote hostname and the login to use").Required().StringVar(&cf.UserHost)
 	ssh.Arg("command", "Command to execute on a remote host").StringsVar(&cf.RemoteCommand)
 	ssh.Flag("port", "SSH port on a remote host").Short('p').Int16Var(&cf.NodePort)
+	ssh.Flag("forward-agent", "Forward agent to target node").Short('A').BoolVar(&cf.ForwardAgent)
```
Comment (klizhentas) [inline tool/tsh/tsh.go]: is the long flag consistent with ssh?

**Review by klizhentas — CHANGES_REQUESTED** [review]
some changes have to be made

File: tool/tsh/tsh.go (hunk 6 above)
Comment (russjones) [reply tool/tsh/tsh.go]: OpenSSH dose not really have a long flag for `-A`. You can do `-o "ForwardAgent yes"` which we are kind of consistent with.

**Review by klizhentas — APPROVED** [review]

**Comment by russjones** [comment]
@gravitational-jenkins retest this please

**Comment by russjones** [comment]
@gravitational-jenkins retest this please

**Pull request merged**